design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+be trusted **only** for the RPC purpose, removing the possibility of abuse.
+To reduce the number of code paths, this process will happen unconditionally,
+even when TLS for Ironic itself is not enabled.
+
+[kea]: https://www.isc.org/kea/
+[issue3]: https://github.com/metal3-io/ironic-standalone-operator/issues/3
+
+#### Architecture FAQ
+
+Q: Why cannot we split dnsmasq into a separate deployment in the non-HA
+architecture? A: That may require having more than one IP address on the
+provisioning network: for dnsmasq and for httpd/ironic. This is a new
+operational requirement that I'd like to avoid at this stage.
+
+Q: Why does the same dnsmasq limitation affect the HA architecture. A: The HA
+architecture is completely new here, so we can introduce new requirements
+without regression in the operational experience.
+
+Q: Why using *DaemonSets* if *StatefulSets* provide us an easier way to address
+separate Ironic instances? A: While we're relying on host networking, making
+several Ironic instances co-exist on the same Kubernetes node is too complex.
+
+Q: Why aren't we using HostPort services? A: The fact that they provide a
+random port is a roadblock for production deployments since many of them
+require opening a predictable port in the firewall configuration. If we use a
+pre-defined port, it may cause conflicts with other HostPort services or even
+end up outside of the allowed range.
+
+#### Current API design
+
+Currently, the API consists of two main objects: `IronicDatabase` and `Ironic`.
+
+The `IronicDatabase` object is very simple:
+
+- `credentialsRef` - a reference to a secret with credentials (generated if
+  missing)
+- `image` - container image to use
+- `tlsRef` - a reference to a TLS secret to use for the service
+
+The `Ironic` object is much more complex and should probably be split into more
+custom resources as we polish its internal architecture. Currently, it uses
+nested structures to logically group fields. Here are the most important fields
+(omitting various fine-tuning for brevity):
+
+- `credentialsRef` - a reference to a secret with credentials (generated if
+  missing)
+- `databaseRef` - a reference to an `IronicDatabase` object (if needed)
+- `distributed` - a boolean flag that enables the HA architecture
+- `networking` - a nested structure that defines networking (see below)
+- `nodeSelector` - a selector for nodes to run Ironic on
+- `tlsRef` - a reference to a TLS secret to use for the service
+
+The `networking` sub-structure deserves a separate consideration:
+
+- `apiPort`, `imageServerPort`, `imageServerTLSPort` allow overriding listening
+  ports for the services
+- `bindInterface` - a boolean flag that makes Ironic listen on only the
+  provisioning interface
+- `dhcp` - another nested structure with DHCP parameters (see below)
+- `externalIP` - IP through which nodes deployed over virtual media access
+  Ironic and HTTPD
+- `interface`, `ipAddress`, `macAddresses` - various ways to specify the
```

Comment:
Similarly, how does provisioningInterface, provisioningIPAddress and provisioningMACAddress sound?

Reply:
There is a potential for confusion here. We call "provisioning network" a *dedicated* network for the PXE and other provisioning traffic. Ironic can work without it, and the variables work regardless of whether you have a "provisioning network".

Maybe I'm overthinking it? Would like more opinions here as well.

Reply:
If we do decide to change this, I'd rather rename the whole "networking" sub-field into something like "provisioningNetwork" to avoid duplicate prefixes.

Reply:
But the rest of the sub-fields seem unrelated to the provisioning network (like dhcp and external ip)?
How would the interface, ipAddress and macAddresses fields work if there is no provisioning interface?
(I am not too familiar with ironic's networking, so just trying to understand the expected behaviour)

Reply:
DHCP is very-very much related to the provisioning network: it's DHCP on the provisioning network. ExternalIP is not directly related indeed, maybe we'd need to move it out then.

> How would the interface, ipAddress and macAddresses fields work if there is no provisioning interface?

There must be a provisioning interface, but it can be on some existing network rather than an isolated provisioning network. Confusing, I know. This is why I'm pondering our usage of "provisioning" here.

Reply:
I would also say to not use "provisioning", on Bare Metal Life Cycle management level the name "Provisioning Nework" makes sense but if we just talk about an ironic instance not that much. Everything Ironic does with the machin goes through the provisioning network, ofc he BMC might have it's own separate network but ipxe, dhcp, bootp,  http image server, IPA - Ironic communication they all happen in the "provisioning" network but from Ironic perspective it is simply "the Network".